modVM/src/org/aion/vm/KernelInterfaceForAVM.java

```diff
@@ -76,26 +80,32 @@ public void putCode(Address address, byte[] code) {
 
     @Override
     public void putObjectGraph(Address contract, byte[] graph) {
-        //Todo: implement it when avm is ready.
+        // TODO: should the vm type be saved
+        this.repositoryCache.saveObjectGraph(contract, graph);
     }
 
     @Override
     public byte[] getObjectGraph(Address contract) {
-        //Todo: implement it when avm is ready.
-        return new byte[0];
+        return this.repositoryCache.getObjectGraph(contract);
     }
 
     @Override
     public void putStorage(Address address, byte[] key, byte[] value) {
         ByteArrayWrapper storageKey = new ByteArrayWrapper(key);
         ByteArrayWrapper storageValue = new ByteArrayWrapper(value);
         this.repositoryCache.addStorageRow(address, storageKey, storageValue);
+        if (this.repositoryCache.getVmType(address) != AVM_CREATE_CODE) {
```

Comment:
the check looks weird (also the L106) What's the reason for forcing the storage change the vm type?

Reply:
Each kernel interface is called only if the contract fit its definition. Whenever the call fits a contract behavior like adding storage, code or the object graph I check if the VM code was not set yet. Otherwise, we cannot ensure that the correct code is set until the blockchain checks for contract transactions. Without this setting, we cannot guarantee that a contract can be correctly deployed and called in the same block.

Reply:
I don't like it either, but so far there is no such method. We need to add a `createContract` to the interfaces that would have to be called whenever a contract is created. This way we can also avoid instantiating and managing contract details objects for all the created accounts that most often are not actual contracts.